packages/swing-store/src/transcriptStore.js

```diff
@@ -235,6 +307,106 @@ export function makeTranscriptStore(
     return bounds;
   }
 
+  const sqlGetSpanEndPos = db.prepare(`
+    SELECT endPos
+    FROM transcriptSpans
+    WHERE vatID = ? AND startPos = ?
+  `);
+  sqlGetSpanEndPos.pluck(true);
+
+  const sqlReadSpanItems = db.prepare(`
+    SELECT item
+    FROM transcriptItems
+    WHERE vatID = ? AND ? <= position AND position < ?
+    ORDER BY position
+  `);
+
+  /**
+   * Read the items in a transcript span
+   *
+   * @param {string} vatID  The vat whose transcript is being read
+   * @param {number} [startPos] A start position identifying the span to be
+   *    read; defaults to the current span, whatever it is
+   *
+   * @returns {IterableIterator<string>}  An iterator over the items in the indicated span
+   */
+  function readSpan(vatID, startPos) {
+    /** @type {number | undefined} */
+    let endPos;
+    if (startPos === undefined) {
+      ({ startPos, endPos } = getCurrentSpanBounds(vatID));
+    } else {
+      insistTranscriptPosition(startPos);
+      endPos = sqlGetSpanEndPos.get(vatID, startPos);
+      if (typeof endPos !== 'number') {
+        throw Fail`no transcript span for ${q(vatID)} at ${q(startPos)}`;
+      }
+    }
+    startPos <= endPos || Fail`${q(startPos)} <= ${q(endPos)}}`;
+    const expectedCount = endPos - startPos;
+
+    function* reader() {
+      let count = 0;
+      const compressedSpanReader = readCompressedSpan(vatID, startPos);
+      if (compressedSpanReader) {
+        for (const item of compressedSpanReader()) {
+          yield item;
+          count += 1;
+        }
+      } else {
+        for (const { item } of sqlReadSpanItems.iterate(
+          vatID,
+          startPos,
+          endPos,
+        )) {
+          yield item;
+          count += 1;
+        }
+      }
+      count === expectedCount ||
+        Fail`read ${q(count)} transcript entries (expected ${q(
+          expectedCount,
+        )})`;
+    }
+
+    if (startPos === endPos) {
+      return empty();
+    }
+
+    return reader();
+  }
+
+  const sqlAddCompressedSpan = db.prepare(`
+    INSERT INTO transcriptCompressedSpans (vatID, startPos, endPos, items, incarnation)
+    VALUES (?, ?, ?, ?, ?)
+  `);
+
+  const sqlDeleteSpanItems = db.prepare(`
+    DELETE FROM transcriptItems
+    WHERE vatID = ? AND position >= ? AND position < ?
+  `);
+
+  function compressSpan(vatID, startPos, endPos, incarnation) {
```

Comment:
This function should take just `vatID, startPos`, and it should look up the other arguments. It will only ever be called on an existing span, so they should always be available, and the lookup will serve as a check that the caller isn't asking for something impossible. And allowing the caller to specify `endPos, incarnation` means they might get it wrong, which we either need to check against the span record anyways, or we allow them to make us create corrupt/unusable compressed records.

So start the function with a query that finds the span record from `vatID, startPos`, asserts that it really exists, and extracts the `endPos, incarnation` from that.

Also, we should guard against `compressSpan` being called twice on the same span (which I think would trigger a SQL `PRIMARY KEY` constraint violation). I'm thinking we query for the compressed record and if it exists, we just exit silently. I'm not sure how likely it would be for someone to make that mistake.. maybe a blind iteration over all spans, compressing all of them? If someone wanted to do that, would they even have a way to avoid making the mistake? I don't think we expose the compression status of anything. So silent-ignore might be the best policy.

Reply:
Not sure I agree.  This would require extra database reads even though it's an internal function called by code that has literally just read the `endPos` and `incarnation` values for purposes of its own.  It's only exposed on the debug API for testing purposes.

Reply:
as discussed, that argument works for me.. I was thinking of it as an external function (perhaps used by a separate tool which compressed spans in the background).

We walked through how it might wind up getting called twice on the same span (maybe if an import dataset contained duplicate artifacts), but we decided that the PRIMARY KEY violation would be an adequate error signal.